C#.md

```diff
@@ -0,0 +1,497 @@
+
+```C#
+// bad
+var value = a && b ? 11 : a ? 10 : b ? 1 : 0;
+
+// good
+var value = isSimple ? 11 : 1;
+```
+
+Ternary expressions should never be nested because they just add to the confusion.
+
+**Centralize duplicate logic in utility functions**
+
+**Using the `dynamic` keyword allowed only in tests**
+
+**Use a StringBuilder instead of string if multiple concatenations are required**
+
+**Override ToString method for the types which you want to provide with custom information.**
+
+**Avoid straightaway copy/pasting of code from other sources. It is always recommended to hand write the code.**
+
+**Usage of 'out' and 'ref' keywords be avoided as recommended by Microsoft [(in the Code analysis Rules and guidelines)](https://msdn.microsoft.com/en-us/library/dd264925.aspx)**
+
+## Best Practices
+
+### Minimize Visibility
+
+In a class API, you should support access to any methods and fields that you make accessible. Therefore, only expose what you intend the caller to use. This can be imperative when writing thread-safe code.
+
+```C#
+public class Parser 
+{
+  // Bad.
+  //   - Callers can directly access and mutate, possibly breaking internal assumptions.
+  public Dictionary<String, String> RawFields;
+
+  // Bad.
+  //   - This is probably intended to be an internal utility function.
+  public String ReadConfigLine() 
+  {
+    ..
+  }
+}
+
+// Good.
+//   - rawFields and the utility function are hidden
+//   - The class is package-private, indicating that it should only be accessed indirectly.
+class Parser 
+{
+  private Dictionary<String, String> m_RawFields;
+
+  private string ReadConfigLine() 
+  {
+    ..
+  }
+}
+```
+
+### Exceptions
+
+**Catch narrow exceptions**
+
+Sometimes when using try/catch blocks, it may be tempting to just `catch Exception`, so you don't have to worry about what type was thrown. This is usually a bad idea, as you can end up catching more than you really wanted to deal with. For example, `catch Exception` would capture `NullReferenceException`, and would capture `OutOfMemoryException`.
+
+```C#
+// Bad.
+//   - If a OutOfMemoryException happens, the program continues rather than aborting.
+try 
+{
+  storage.InsertUser(user);
+} 
+catch (Exception ex) 
+{
+  LOG.Error("Failed to insert user.");
+}
+
+try 
+{
+  storage.InsertUser(user);
+} 
+catch (StorageException ex) 
+{
+  LOG.Error("Failed to insert user.");
+}
+```
+
+**Don't swallow exceptions**
+
+An empty catch block is usually a bad idea, as you have no signal of a problem. Coupled with narrow exception violations, it's a recipe for disaster.
+
+**Throw appropriate exception types**
+
+Let your API users obey the "catch narrow exceptions" rule, don't throw `Exception`. Even if you are calling another naughty API that throws `Exception`, at least hide that so it doesn't bubble up even further. You should also make an effort to hide implementation details from your callers when it comes to exceptions.

```

Comment:
It makes sense to link the "method comments" section of the guidelines somewhere here.